pytorch_memlab/line_profiler.py

```diff
@@ -58,128 +204,143 @@ def add_function(self, func):
 
-    def trace_callback(self, frame, event, arg):
+    def _trace_callback(self, frame, event, arg):
         """Trace the execution of python line-by-line"""
 
         if event == 'call':
-            return self.trace_callback
+            return self._trace_callback
 
-        if event in ['line', 'return'] and frame.f_code in self.code_map:
-            line_stat = self.code_map[frame.f_code]['line_stat']
+        code_hash = hash(frame.f_code)
+        if event in ['line', 'return'] and code_hash in self._code_infos:
+            code_info = self._code_infos[code_hash]
             with torch.cuda.device(self.target_gpu):
-                allocated_memory = torch.cuda.memory_allocated()
-                cached_memory = torch.cuda.memory_cached()
-                torch.cuda.empty_cache()
-            if event == 'return':
-                lineno = max(line_stat.keys()) + 1
+                self._raw_line_records.append({
+                    'code_hash': code_hash, 
+                    'line': code_info['prev_line'],
+                    'prev_record_idx': code_info['prev_record'],
+                    **torch.cuda.memory_stats()})
+                self._reset_cuda_stats()
+
+            if event == 'line':
+                code_info['prev_line'] = frame.f_lineno
+                code_info['prev_record'] = len(self._raw_line_records)-1
+            elif event == 'return':
+                code_info['prev_line'] = code_info['first_line']
+                code_info['prev_record'] = -1
             else:
-                lineno = frame.f_lineno
-            last_lineno = self.code_map[frame.f_code]['last_lineno']
-            line_stat[last_lineno].append((allocated_memory, cached_memory))
-            self.code_map[frame.f_code]['last_lineno'] = lineno
-        return
-
-    def print_stats(self, stream=None):
-        """Print the stat of each functions
+                assert False
+
+    def line_records(self, func=None, columns=DEFAULT_COLUMNS):
+        """Returns a (line, statistic)-indexed dataframe of memory stats.
+        
+        The columns are explained in the PyTorch documentation:
+        
+        https://pytorch.org/docs/stable/cuda.html#torch.cuda.memory_stats
         """
-        for code, stat in self.code_map.items():
-            show_func(
-                filename=code.co_filename,
-                trace_stat=stat,
-                stream=stream
-            )
+        if len(self._raw_line_records) == 0:
+            return pd.DataFrame(index=pd.MultiIndex.from_product([[], []]), columns=columns)
 
-    def print_func_stats(self, func, stream=None):
-        """Print the stat of a registered function"""
-        code = func.__code__
-        if code in self.code_map:
-            show_func(
-                filename=code.co_filename,
-                trace_stat=self.code_map[code],
-                stream=stream
-            )
+        line_records = _line_records(self._raw_line_records, self._code_infos)
```

Comment:
Can we call `_line_records` in `_subset_line_records` to hide the implementation details.

Reply:
Not without passing raw_line_records and code_infos to it too, and so confusing intent.